lgc/patch/PatchReadFirstLane.cpp

```diff
@@ -154,6 +155,87 @@ void PatchReadFirstLane::getAnalysisUsage(AnalysisUsage &analysisUsage) const {
   analysisUsage.setPreservesCFG();
 }
 
+// =====================================================================================================================
+// Use the uniform (non-divergent) node instead of a divergent one if they are equal.
+//
+// Detect a case where a branch condition tests two nodes for equality where one of them
+// is divergent and the other uniform. In the true block replace the divergent
+// use with the uniform one as it helps instruction selection decide what to put in an SGPR.
+//
+// For example, in the code snippet below in BB1, %divergent can be replaced with %uniform.
+//
+// BB0:
+// %uniform = call i32 @llvm.amdgcn.readlane(i32 %divergent, i32 %lane)
+// %cmp = icmp eq i32 %divergent, %uniform
+// br i1 %cmp, label %BB1, label %BB2
+//
+// BB1:
+// ..
+// %use = zext i32 %divergent to i64 ; can use %uniform
+//
+// @param [in,out] function : LLVM function to be run for the optimization.
+bool PatchReadFirstLane::promoteEqualUniformOps(Function &function) {
+
+  bool changed = false;
+  for (auto &bb : function) {
+    // Look for a block that has a single predecessor with conditional branch.
+    const auto *pred = bb.getSinglePredecessor();
+    if (!pred)
+      continue;
+    const auto *branchInst = dyn_cast<BranchInst>(pred->getTerminator());
+    if (!branchInst || !branchInst->isConditional())
+      continue;
+    assert(branchInst->getSuccessor(0) != branchInst->getSuccessor(1));
+
+    // Make sure the condition for entering bb is the equality test passing.
+    bool match = false;
+    const auto *cmpInst = dyn_cast<ICmpInst>(branchInst->getCondition());
+    if (cmpInst) {
+      switch (cmpInst->getPredicate()) {
+      case ICmpInst::ICMP_EQ:
+        match = branchInst->getSuccessor(0) == &bb;
+        break;
+      case ICmpInst::ICMP_NE:
+        match = branchInst->getSuccessor(1) == &bb;
+        break;
+      default:
+        break;
+      }
+    }
+    if (!match)
+      continue;
+
+    // Check if the condition compares a divergent and uniform value.
+    Instruction *divergentInst = nullptr;
+    Value *uniformVal = nullptr;
+    for (int i = 0; i < 2; ++i) {
+      const Use &cmpArg = cmpInst->getOperandUse(i);
+      if (m_divergenceAnalysis->isDivergentUse(&cmpArg))
+        divergentInst = dyn_cast<Instruction>(&cmpArg);
+      else
+        uniformVal = dyn_cast<Value>(&cmpArg);
```

Comment:
You shouldn't need any cast here -- that's the benefit of making it a `Value *`.